crates/driver/src/domain/competition/bad_tokens/mod.rs

```diff
@@ -0,0 +1,121 @@
+    pub fn with_config(mut self, config: HashMap<eth::TokenAddress, Quality>) -> Self {
+        self.hardcoded = config;
+        self
+    }
+
+    /// Enables detection of unsupported tokens via simulation based detection
+    /// methods.
+    pub fn with_simulation_detector(&mut self, detector: simulation::Detector) -> &mut Self {
+        self.simulation_detector = Some(detector);
+        self
+    }
+
+    /// Enables detection of unsupported tokens based on heuristics.
+    pub fn with_heuristic_detector(&mut self) -> &mut Self {
+        self.metrics = Some(metrics::Detector);
+        self
+    }
+
+    /// Removes all unsupported orders from the auction.
+    pub async fn filter_unsupported_orders_in_auction(&self, mut auction: Auction) -> Auction {
+        let now = Instant::now();
+
+        let filtered_orders = futures::stream::iter(auction.orders.into_iter())
+            .filter_map(move |order| async move {
+                let sell = self.get_token_quality(order.sell.token, now);
+                let buy = self.get_token_quality(order.sell.token, now);
+                match (sell, buy) {
+                    // both tokens supported => keep order
+                    (Some(Quality::Supported), Some(Quality::Supported)) => Some(order),
+                    // at least 1 token unsupported => drop order
+                    (Some(Quality::Unsupported), _) | (_, Some(Quality::Unsupported)) => None,
+                    // sell token quality is unknown => keep order if token is supported
+                    (None, _) => {
+                        let Some(detector) = &self.simulation_detector else {
+                            // we can't determine quality => assume order is good
+                            return Some(order);
+                        };
+                        let quality = detector.determine_sell_token_quality(&order, now).await;
+                        (quality == Some(Quality::Supported)).then_some(order)
+                    }
+                    // buy token quality is unknown => keep order (because we can't
+                    // determine quality and assume it's good)
+                    (_, None) => Some(order),
+                }
+            })
+            .collect::<Vec<_>>()
+            .await;
+        auction.orders = filtered_orders;
+
+        if let Some(detector) = &self.simulation_detector {
+            detector.evict_outdated_entries();
+        }
+
+        auction
+    }
+
+    fn get_token_quality(&self, token: eth::TokenAddress, now: Instant) -> Option<Quality> {
+        if let Some(quality) = self.hardcoded.get(&token) {
+            return Some(*quality);
+        }
+
+        if let Some(detector) = &self.simulation_detector {
+            if let Some(quality) = detector.get_quality(token, now) {
+                return Some(quality);
+            }
+        }
+
+        if let Some(metrics) = &self.metrics {
+            return metrics.get_quality(token);
+        }
```

Comment:
The idea was to have multiple priorities:
1. hardcoded list
2. simulation based
3. metrics based

The reason to give simulation based detection a higher prio is that it should have very few false negatives (especially with the new idea of simulating from the user's address). That way we don't even let the token enter the system to collect metrics on it for the metric based solution to detect anything.

> Maybe, metrics detector needs an instance of the simulation detector, which will be called internally

These are 2 completely different strategies to detect bad tokens.

> and instead of 2 detectors, there will be only one(e.g. InstrumentedDetector)?

This wrapper type is currently `bad_tokens::Detector`. I wanted to avoid introducing abstractions too early to see which interfaces actually make sense. Also given that these different strategies act on very different data I'm not sure building a trait around that would be super helpful anyway.

Reply:
> That way we don't even let the token enter the system to collect metrics on it for the metric based solution to detect anything.

Got it, totally makes sense.

Reply:
OTOH, thinking more about the metrics detector, it should probably still require access to the simulating detector in case of "Good" quality to reset its statistic for the token. Otherwise, metrics detector never knows whether the token was considered "Good" at all and will be incrementing its "Bad" token blindly always returning the same result.